lib/include/jxl/codestream_header.h

```diff
@@ -314,6 +314,32 @@ typedef struct {
   /** Indicates this is the last animation frame.
    */
   JXL_BOOL is_last;
+
+  /** If coalescing is enabled, the following can be ignored; crop_x0/crop_y0 is
+     always zero, xsize/ysize is always equal to the image dimensions, and
+     blendmode is always 0 (kReplace). If coalescing is disabled, the layer has
+     dimensions xsize x ysize (which can be smaller or larger than the image
+     dimensions) and its top left corner is at position (crop_x0,crop_y0) -
+     these offsets can be negative.
+   */
+  /** Horizontal offset of the frame (can be negative)
+   */
+  int32_t crop_x0;
+  /** Vertical offset of the frame (can be negative)
+   */
+  int32_t crop_y0;
+  /** Width of the frame (number of columns)
+   */
+  uint32_t xsize;
+  /** Height of the frame (number of rows)
+   */
+  uint32_t ysize;
+  /** Blend modes: 0=kReplace, 1=kAdd, 2=kBlend, 3=kMulAdd, 4=kMul
+   */
+  uint32_t blendmode;
```

Comment:
this one would make sense as an enum

Reply:
Agreed, or a struct with an enum in it, depending on how much we want to expose. If the goal is to expose enough info to let the application do all the blending itself, then we also need to expose source, clamp, alpha channel, and blendmodes for extra channels - but I'm not sure if it makes sense to expose all that; it might be better to make no-coalescing only work for "simple" layers (the kind of layers you could get with GIF/APNG/XCF/PSD as input), and make it return a decode error if too exotic stuff is happening.

Reply:
I don't see why this would need to be a partial implementation of the feature. What do you do if coalescence is false but the .jxl file contains non-"simple" layers? fail when the coalescing version would have succeeded?? That's very odd behavior.

Reply:
Yes, we could expose all relevant FrameHeader fields (`blending_info`, `ec_blending_info`, and `save_as_reference`), and then in theory an application could manually keep track of the 3 reference slots (the fourth one can only be used by Patches so wouldn't be exposed) and do all the blending. Maybe that's a better approach.

I was just assuming that in most applications, the only thing they will care about is RGBA with kReplace or kBlend blending (as that is what is needed for transcoding GIF/APNG/XCF/PSD, and what is already implemented in image libraries like ImageMagick). So there's a risk that they will only implement that. From a practical point of view, it doesn't make a big difference if the application fails because libjxl returns a not_supported error, or if it fails because libjxl returns the actual exotic blendinfo and then the application gives up.

The advantage of returning the full blendinfo is that it leaves open the possibility for applications to actually implement exotic blending itself, as well as leaving open the possibility to extract layers even if the application does not know how to blend them (which could be useful e.g. for a jxltran application that can convert a lossless jxl to a lossy jxl without making all frames canvas-sized in the process).

The disadvantage of returning the full blendinfo is that it's quite a bit of extra stuff to add to the API (not a big issue if we can put it behind a compile flag), and that there's a risk that applications will not actually inspect all the details of how the blending should be done, make assumptions instead, and produce incorrect results (which would of course be an application bug, not a libjxl bug, but still...).

I don't have a strong opinion either way; I'm leaning towards returning the full blendinfo but I would also be fine returning a "not supported" error when there is exotic blendinfo and exposing only a simplified subset of the info that is easier for applications to deal with (and then of course document well that setting coalescing to false can have that consequence).

Reply:
Updated this part. It is now a full implementation (exposes all blending info).